### ISSUE TITLE
[RFC] OpenSC should allow to sign data with length more than 512 bytes.

### ISSUE THREAD

**Issue by ya-isakov** [OpenSC/OpenSC#2300]
As a continuation of discussion in #2293, I've found that at least Yubikey is (almost certanly) supports ed25519 without any prehashing possible. This should work for small data, but won't work for e.g. X509 cettficates, which could be really longer than 512 bytes.

From RFC 8410, there are only OIDs for eddsa in pure mode, and no other hashing is defined, so we cannot prehash certificates ourselves.

I've managed to sign X509 using Yubikey and XCA with my changes, and openssl verifies this certificate, so I could assume that my understanding is correct.

So, one proposed solution is to change length of two buffers in OpenSC to be, for exampe, 2048 bytes, which should fit most X509 certificates, and still fits into max size of APDU

**Comment by ya-isakov** [comment]
@dengert, continuing our discussion, I think that there is no point to do a prehashing in e.g. OpenSC - I mean, from what I've read in RFC8032, it's not possible to get PureEdDSA signature from ed25519ph, so we cannot do message prehash in OpenSC, put hash in card to sign, and then, somehow, pretend it to be a PureEdDSA hash.
Also, as I described in #2287, I cannot find a way to specify in OpenPGP APDUs, what type of ed25519 we want, and gnupg is using PureEdDSA on card as well. So, I'm assuming that there is no way, currently, to provide prehashed version of ed25519 in PKCS#11 interface of OpenSC.

What do you think?

**Comment by ya-isakov** [comment]
Continuing from #2293
> Certificates to be signed are a good example, but there should be no limit. A user could ask for a large document to be signed. https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/csprd01/pkcs11-curr-v3.0-csprd01.html has:
> 
> > Table 44, EdDSA: Key and Data Length
> > Function | Key type | Input length | Output length
> > -- | -- | -- | --
> > C_Sign   | CKK_EC_EDWARDS private key | any                  | 2bLen
> > C_Verify | CKK_EC_EDWARDS public key   | any, £2bLen 2 | N/A
> 
> > Note that for EdDSA in pure mode, Ed25519 and Ed448 the data must be processed twice. Therefore, a token might need to cache all the data, especially when used with C_SignUpdate/C_VerifyUpdate. If tokens are unable to do so they can return CKM_TOKEN_RESOURCE_EXCEEDED.
> 
> So I was suggesting that the card driver could collect up multiple segment of data much like a driver doing the hash in software but not doing the hash, which was why I brought up `card-sc-hsm.c` as example of how a driver could accept any size data then use extended APDUs or command chaining to the card, or do some of the processing in software so the card does not have to look at the data twice. But the "the data must be processed twice" may be a limiting factor.

The problem with it, that jusdging from OpenPGP spec, sign should be done in one operation. I've tried to use multipe S_SignUpdate, and it seems that all of them are writing to the same buffer, and this buffer is sent in one APDU to a card. Yubikey supports Extended Length, so I can fit ~3k bytes into it.

I cannot find any limitations in PKCS#11 standard, about maximum length of data to be signed. From my point of view, OpenSC should not limit it artificially. If card cannot send large data, it will return error, which will be returned as Generic Error in PKCS#11.
> 
> An trace of the Yubikey at the pcscd level would show what the the Yubkey driver or OpenPGP is doing.
> 
> What I am saying is I don't think changing the size of the buffer in pkcs15-sec.c is the correct way to address this.
> 
> PKCS11 implies there might be some token resource limit to what can be handled.
> @Jakuje's test with the NitroKey may have a 255 byte limit, or maybe the extended APDUs might allow a larger limit.
> 
> If all the tokens have some reasonable limit, like 4K, then maybe pkcs15-sec.c buffer size would be OK.
> 
> P.S. I found I do have a NitroKey Start token that supports Ed25519. I hope to be able to do some testing with it soon.

**Comment by ya-isakov** [comment]
Actually, it's really strange, that S_Sign supports any input, if it should be hashed, but for mechanisms without hashing, it supports only 512 bytes of data.

**Comment by ya-isakov** [comment]
@dengert Actually, I think that drivers you mentioned, will have the same limitation of 512 bytes for e.g CKM_ECDSA (without hashing), as it's a global, and do not depend on driver used.

**Comment by ya-isakov** [comment]
Let me summarize all my messages a little bit.

- In OpenSC, there is a buffer for not-hashed mechanisms like CKM_EDDSA/CKM_ECDSA for C_Sign/C_SignUpdate with maximum size of 512 bytes. Even if I'll do multiple C_SIgnUpdate with smaller data chunks, the total data cannot be more than 512 bytes.
- To fit more than 512 bytes, CKM_ECDSA allows to use mechanisms with hashing, which will update hash and send hash to a card. OpenPGP has no means (in the standard) to specify such hashing, so card knows only that some data needs to be hashed.
This works for ECDSA, as e.g. X509 defines ecdsa-with-sha256 as sha256 hashing of input and use of this value for signing.
- There are two versions of ed25519 - PureEdDSA, which hashes data internally, and HashEdDSA, which allows some prehashing. These two variants are incompatible between each other, as in HashEdDSA some internal flag (phflag) must be set to 1. This parameter is a part of implementation of EdDSA, so it must be supported by token, and token should have mechanism to specify it. Without setting this flag, token will use PureEdDSA for signing, so prehashing of data and sending it to token will not give you HashEdDSA signature
- HashEdDSA is not recommended at all, according to RFC8032 - `prehashing also makes the functions greatly more vulnerable to weaknesses in hash functions used.  These variants SHOULD NOT be used`
- X509 supports only PureEdDSA variant of EDDSA, so, tokens will (and are) working with it. Certificates could be longer than 512 bytes, and they cannot be signed by current OpenSC, because of the limits on buffers.

**Comment by dengert** [comment]
> * In OpenSC, there is a buffer for not-hashed mechanisms like CKM_EDDSA/CKM_ECDSA for C_Sign/C_SignUpdate with maximum size of 512 bytes. Even if I'll do multiple C_SIgnUpdate with smaller data chunks, the total data cannot be more than 512 bytes.

The limits in OpenSC are based on what was needed when the code was written. Changing the size of the buffer to say 4K would be another such limit. What I was trying to get at, is there another way to pass in to the card drive whatever the C_SignUpdate sends. Maybe the buffer should not be static, but use `realloc`.  I believe that this is what the the card-sc-hsm.c is doing, it receives chucks of data, and and passes to card to do the hash on the card, or does part of the hash in software, and the last chunk on the card. 

> * To fit more than 512 bytes, CKM_ECDSA allows to use mechanisms with hashing, which will update hash and send hash to a card. OpenPGP has no means (in the standard) to specify such hashing, so card knows only that some data needs to be hashed. This works for ECDSA, as e.g. X509 defines ecdsa-with-sha256 as sha256 hashing of input and use of this value for signing.

Yes, all the other sign algorithms have small limits of the size of data that can be signed and use hashing to sign the hash rather then the actual data.  

> * There are two versions of ed25519 - PureEdDSA, which hashes data internally, and HashEdDSA, which allows some prehashing. These two variants are incompatible between each other, as in HashEdDSA some internal flag (phflag) must be set to 1. This parameter is a part of implementation of EdDSA, so it must be supported by token, and token should have mechanism to specify it. Without setting this flag, token will use PureEdDSA for signing, so prehashing of data and sending it to token will not give you HashEdDSA signature

Looks like the prehashing was a way to make EdDSA more like the older sign algorithms, i.e. limiting the size of the input to the size of the prehash, thus allowing unlimited size data to be signed.  
 
> * HashEdDSA is not recommended at all, according to RFC8032 - `prehashing also makes the functions greatly more vulnerable to weaknesses in hash functions used.  These variants SHOULD NOT be used`

The problem appears to be https://tools.ietf.org/html/rfc8032 "5.1.6.  Sign" "The inputs to the signing procedure is the private key, a 32-octet  string, and a message M of arbitrary size" and "Compute SHA-512(dom2(F, C) || prefix || PH(M))" where PH(M)=M.  Thus the hash has to be done on the token to not  expose the prefix. But hashing algorithms are designed to handle unlimited data, in chucks. 

My speculation is OpenPGP and/or the current cards don't know how to pass the M in chunks to the card but require it in one chuck. As pointed out the Yubikey can accept some 3000+ bytes, where as the NitroKey appears to be limited to 255. 

I could be wrong, and we just don't know how to pass the M in chunks. ISO7816 has ways to tell a token to do a hash in chucks, but not clear how to tell the the token to accept the M in chucks and signal the last chuck. (Maybe command chaining?) but does the applet implement the ISO7816 to know how to handle it, or does the applet buffer up the chucks in order to do the hash in one chuck. 

PKCS#11  has the CKM_TOKEN_RESOURCE_EXCEEDED, so there is someway to indicate the M is too long. 
        
> * Probably because of this recommendation, X509 supports only PureEdDSA variant of EDDSA, so, tokens will (and are) working for signing X509 certs. Certificates could be longer than 512 bytes, and they cannot be signed by current OpenSC, because of the limits on buffers.

So in the short term we could change the pkcs15-sec.c 512 byte buffer to something like 4K, or make it dynamic and use `realloc` or pass the chunks to the card driver so it appends the chucks to send as one large chuck. But eventually will need to address the problem of sending the M to a token in chunks. 

P.S. I found I have a Yubikey 5 NFC with firmware 5.2.6 that can handle EdDSA. I have this and the Nitro loaded with EdDSA keys, and have been trying some tests using "gpg (GnuPG) 2.2.19" to look at APDUs.
 So does https://github.com/ya-isakov/OpenSC/tree/ed25519 contain everything I would need to do some testing with OpenSC?

**Comment by ya-isakov** [comment]
> > * In OpenSC, there is a buffer for not-hashed mechanisms like CKM_EDDSA/CKM_ECDSA for C_Sign/C_SignUpdate with maximum size of 512 bytes. Even if I'll do multiple C_SIgnUpdate with smaller data chunks, the total data cannot be more than 512 bytes.
> 
> The limits in OpenSC are based on what was needed when the code was written. Changing the size of the buffer to say 4K would be another such limit. What I was trying to get at, is there another way to pass in to the card drive whatever the C_SignUpdate sends. Maybe the buffer should not be static, but use `realloc`. I believe that this is what the the card-sc-hsm.c is doing, it receives chucks of data, and and passes to card to do the hash on the card, or does part of the hash in software, and the last chunk on the card.

I thought about using dynamic memory, but this has to be allocated in C_SignInit, reallocated on each C_SIgnUpdate, and freed on C_SignFInal. What would happen, if something happened before C_SIgnFinal will be called? We will get a memory leak.

The problem is that this is done long before card-sc-hsm starts handling it, as it's called only when C_Sign or C_SignFinal done their work, and these functions will fail on data more than 512 bytes.

> 
> > * To fit more than 512 bytes, CKM_ECDSA allows to use mechanisms with hashing, which will update hash and send hash to a card. OpenPGP has no means (in the standard) to specify such hashing, so card knows only that some data needs to be hashed. This works for ECDSA, as e.g. X509 defines ecdsa-with-sha256 as sha256 hashing of input and use of this value for signing.
> 
> Yes, all the other sign algorithms have small limits of the size of data that can be signed and use hashing to sign the hash rather then the actual data.
> 
> > * There are two versions of ed25519 - PureEdDSA, which hashes data internally, and HashEdDSA, which allows some prehashing. These two variants are incompatible between each other, as in HashEdDSA some internal flag (phflag) must be set to 1. This parameter is a part of implementation of EdDSA, so it must be supported by token, and token should have mechanism to specify it. Without setting this flag, token will use PureEdDSA for signing, so prehashing of data and sending it to token will not give you HashEdDSA signature
> 
> Looks like the prehashing was a way to make EdDSA more like the older sign algorithms, i.e. limiting the size of the input to the size of the prehash, thus allowing unlimited size data to be signed.
even PureEdDSA can sign any amount of data, because it's hashing it itself. Prehashing could be good in case if token would support ed25519ph, and OpenSC is used by app which also works with ed25519ph. This way, 512bytes limit will not affect such app. But, it seems that nothing in real world supports ed25519ph.
> 
> > * HashEdDSA is not recommended at all, according to RFC8032 - `prehashing also makes the functions greatly more vulnerable to weaknesses in hash functions used.  These variants SHOULD NOT be used`
> 
> The problem appears to be https://tools.ietf.org/html/rfc8032 "5.1.6. Sign" "The inputs to the signing procedure is the private key, a 32-octet string, and a message M of arbitrary size" and "Compute SHA-512(dom2(F, C) || prefix || PH(M))" where PH(M)=M. Thus the hash has to be done on the token to not expose the prefix. But hashing algorithms are designed to handle unlimited data, in chucks.
Yes, exactly, SHA512 part has to be done in token, even if PH(M)=M. This is how PureEdDSA can handle any message, unlike e.g. ECDSA.
> 
> My speculation is OpenPGP and/or the current cards don't know how to pass the M in chunks to the card but require it in one chuck. As pointed out the Yubikey can accept some 3000+ bytes, where as the NitroKey appears to be limited to 255.
Yes, exactly. OpenPGP spec specify only one command to do signing. Command is not limited in length, in the spec, so Yubikey allows to sign ~3k bytes using Extended Length ATRs, and Nitrokey, as not supporting these kind of ATRs, should support chaining (If I understand correctly, from the spec)
> 
> I could be wrong, and we just don't know how to pass the M in chunks. ISO7816 has ways to tell a token to do a hash in chucks, but not clear how to tell the the token to accept the M in chucks and signal the last chuck. (Maybe command chaining?) but does the applet implement the ISO7816 to know how to handle it, or does the applet buffer up the chucks in order to do the hash in one chuck.
Command chaining is not recommended by spec, and is called obsolete, in case if Extended Length is supported (which is a case for Yubikeys with firmware 5.2+)
> 
> PKCS#11 has the CKM_TOKEN_RESOURCE_EXCEEDED, so there is someway to indicate the M is too long.
> 
> > * Probably because of this recommendation, X509 supports only PureEdDSA variant of EDDSA, so, tokens will (and are) working for signing X509 certs. Certificates could be longer than 512 bytes, and they cannot be signed by current OpenSC, because of the limits on buffers.
> 
> So in the short term we could change the pkcs15-sec.c 512 byte buffer to something like 4K, or make it dynamic and use `realloc` or pass the chunks to the card driver so it appends the chucks to send as one large chuck. But eventually will need to address the problem of sending the M to a token in chunks.
Buffers in pkcs15-sec and mechanism are needed to be increased, anyway, as no chunk is sent to card driver, until C_Sign/C_SignFinish has all data we want to sign, and these functions are sending this data in one chunk, to card driver, if I understand correctly.
> 
> P.S. I found I have a Yubikey 5 NFC with firmware 5.2.6 that can handle EdDSA. I have this and the Nitro loaded with EdDSA keys, and have been trying some tests using "gpg (GnuPG) 2.2.19" to look at APDUs.
> So does https://github.com/ya-isakov/OpenSC/tree/ed25519 contain everything I would need to do some testing with OpenSC?

This, and also master branch of my repo (so, #2293 and #2286)

**Comment by dengert** [comment]
@ya-isakov   As you pointed out is a private e-mail, https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-10 which is an IETF draft that expired "4 March 2021", says OpenPGP always uses SHA256 of the data with Ed25519 signatures.   

I would like to point out that OpenSC primary function is to provide a PKCS#11 interface to tokens. EdDSA is defined in https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html

The OpenSC card-openpgp.c driver enforces the openpgp crypto of using SHA256 on the input, but needs a default of supporting PKCS\#11 standard. Like ssh appears to do based on you comment:

> P.S. Funny thing, that when gpg-agent is used for ssh, no prehashing is done:
> send apdu: c=00 i=88 p1=00 p2=00 lc=146 le=256 em=0  

The openpgp-tool.c should enforce the OpenPGP  standards for OpenPGP signatures
and do SHA256 and pass to card-openpgp.c for use with OpenPGP internal signatures.  

There will be other tokens (applets) that support EdDSA, and OpenSC needs to be consistent in following the standards. (The two tokens appear to do Ed25519 on the token, in that they appear to have limits on how much data they can accept. Nitro Start accepts 255, Yubikey accepts 3000+.   PKCS\#11 recognizes that tokens may have limited memory and can return CKR_TOKEN_RESOURCE_EXCEEDED.
(The Nitro may accept more if extended APDUs were used. Needs to be tested. )

The main problem is EdDSA is the first asymmetric algorithm supported by OpenSC that can accept
unlimited data which needs to be sent to the token. 

This also points out that GnuPG (gpg) should not be used to sign things like X509 certificates. 

For now, OpenSC could use a 4K, 8K, 16K i.e. something reasonable buffer in pkcs15-sec.c
and if data is larger then  return CKR_TOKEN_RESOURCE_EXCEEDED.

What I have not looked at (but I believe you have) is X509 signatures and Ed25519.
I assume these follow the standards for crypto, unlike OpenPGP that says always
use SHA256.

**Comment by dengert** [comment]
> P.S. Funny thing, that when gpg-agent is used for ssh, no prehashing is done:
send apdu: c=00 i=88 p1=00 p2=00 lc=146 le=256 em=0

That is an INTERNAL AUTHENTICATE 88 operation. PERFORM SECURITY OPERATION (PSO) is 2A. 
OpenPGP VC3.4.1 "7.1 Usage of ISO Standard Commands".

 It is not clear if ssh created the APDU for the gpg-agent to issue, or the gpg-agent ceated the APDU.

**Comment by ya-isakov** [comment]
> @ya-isakov As you pointed out is a private e-mail, https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-10 which is an IETF draft that expired "4 March 2021", says OpenPGP always uses SHA256 of the data with Ed25519 signatures.
> 
> I would like to point out that OpenSC primary function is to provide a PKCS#11 interface to tokens. EdDSA is defined in https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/os/pkcs11-curr-v3.0-os.html
> 
> The OpenSC card-openpgp.c driver enforces the openpgp crypto of using SHA256 on the input, but needs a default of supporting PKCS#11 standard. Like ssh appears to do based on you comment:

No, card-openpgp is not using sha256 hashing at all, it pass the same data as was passed to C_Sign from PKCS#11 app.
> 
> > P.S. Funny thing, that when gpg-agent is used for ssh, no prehashing is done:
> > send apdu: c=00 i=88 p1=00 p2=00 lc=146 le=256 em=0
> 
> The openpgp-tool.c should enforce the OpenPGP standards for OpenPGP signatures
> and do SHA256 and pass to card-openpgp.c for use with OpenPGP internal signatures.

I think, there could be some mixing of OpenPGP and OpenPGP Card there. They're separate standards, and should not be mixed. card-openpgp.c is OpenSC driver for OpenPGP Card - compatible tokens (which is extension if ISO7816, IIUC). OpenSC could have ability to generate OpenPGP-compatible formats, like armoring etc, but it's completely separate task, and is probably out of scope for OpenSC (as it's implementing PKCS#11, and not RFC4880)
> 
> There will be other tokens (applets) that support EdDSA, and OpenSC needs to be consistent in following the standards. (The two tokens appear to do Ed25519 on the token, in that they appear to have limits on how much data they can accept. Nitro Start accepts 255, Yubikey accepts 3000+. PKCS#11 recognizes that tokens may have limited memory and can return CKR_TOKEN_RESOURCE_EXCEEDED.
> (The Nitro may accept more if extended APDUs were used. Needs to be tested. )
> 
> The main problem is EdDSA is the first asymmetric algorithm supported by OpenSC that can accept
> unlimited data which needs to be sent to the token.
> 
> This also points out that GnuPG (gpg) should not be used to sign things like X509 certificates.

GnuPG is not able to sign X509, at least X509-compatible way. It can sign only in RFC4880-compatible way. There is gpgsm, though, but release version of gnupg doesn't support ed25519 for gpgsm, yet.
> 
> For now, OpenSC could use a 4K, 8K, 16K i.e. something reasonable buffer in pkcs15-sec.c
> and if data is larger then return CKR_TOKEN_RESOURCE_EXCEEDED.
> 
> What I have not looked at (but I believe you have) is X509 signatures and Ed25519.
> I assume these follow the standards for crypto, unlike OpenPGP that says always
> use SHA256.

Yes, after #2293, OpenSC returns RFC8410-compatible public keys, and I verified that it can be used to sign X509 certificates in RFC8410-compatible way (and even verified by OpenSSL)

**Comment by dengert** [comment]
>For performance reasons chaining should be avoided, all modern card readers support APDUs with 2 KB and more.

Command chaining vs extended APDU depends on the card OS, the reader and the applet. And if the reader is limited in its extended APDU (like 2KB), both Extended APDU and chaining could be used. 

There are many people writing applets for various cards and some specs say nothing about supporting extended APDU. OpenSC tries to handle both as best it can. 

Yes, OpenSC is supporting PKCS\#11, to use any token that supports keys, private keys, public keys, and even without X509.certificates. Even when the token can do much more.

**Comment by ya-isakov** [comment]
> > P.S. Funny thing, that when gpg-agent is used for ssh, no prehashing is done:
> > send apdu: c=00 i=88 p1=00 p2=00 lc=146 le=256 em=0
> 
> That is an INTERNAL AUTHENTICATE 88 operation. PERFORM SECURITY OPERATION (PSO) is 2A.
> OpenPGP VC3.4.1 "7.1 Usage of ISO Standard Commands".
> 
> It is not clear if ssh created the APDU for the gpg-agent to issue, or the gpg-agent ceated the APDU.

SSH has no idea about APDUs, ad gnupg is acting as ssh-agent. So, ssh is passing some challenge for "agent" to sign, and gnupg is signing it by Auth key using DSO:INTERNAL AUTHENTICATE

**Comment by dengert** [comment]
> No, card-openpgp is not using sha256 hashing at all, it pass the same data as was passed to C_Sign from PKCS#11 app. Here is a part of my OpenSC log, to proof

Ok, good to hear that is working. The GnuPG is the one enforcing the SHA256 as default as that is what the OpenPGP specs say to do for OpenPGP signatures. 

Next week I hope to your  branch using the Nitro Start and Yubikey 5 NFC which were initialized using gpg to test EdDSA.

**Comment by ya-isakov** [comment]
@dengert Yes, please, my code definitely needs more testing :) I can provide a patch for XCA (which is for testing only, I do not want to publish it, at least until I'll make it better), to test X509 signing (as libp11-based OpenSSL engine is not supporting ed25519, yet)

**Comment by ya-isakov** [comment]
@Jakuje If I understand correctly, Nitrokeys has Gnuk applet on them. I've found this [check](https://github.com/ggkitsas/gnuk/blob/master/src/openpgp.c#L825) in Gnuk , which essentially limits data to be signed to 256 bytes (and cause error you posted in https://github.com/OpenSC/OpenSC/pull/2293#issuecomment-818617537). Funny thing that it is named EDDSA_HASH_LEN_MAX, but Gnuk is doing PureEdDSA, and not HashEdDSA (according to [eddsa_sign_25519](https://github.com/ggkitsas/gnuk/blob/master/src/ecc-edwards.c#L664))

Unfortunately, I have no Gnuk-based tokens (especially where I can upload my own applets) to check if this limit could be lowered. My guess would be that if this limit is raised, command chaining should work and allow to sign more than 256 bytes.

**Comment by dengert** [comment]
The [check](https://github.com/ggkitsas/gnuk/blob/master/src/openpgp.c#L825) is for the OpenPGP appliction on the host. Hopefully, the applet on the token does not have such a small limit.

**Comment by ya-isakov** [comment]
@dengert No, these are sources of applet on card:
```
Gnuk supports OpenPGP card protocol version 3, and it runs on STM32F103 processor.
```
And this limit causes the exact same error, as @Jakuje saw - https://github.com/ggkitsas/gnuk/blob/master/src/openpgp.c#L906

**Comment by ya-isakov** [comment]
My bad, I've used a wrong source, here is official source of Gnuk applet, but it's the same https://salsa.debian.org/gnuk-team/gnuk/gnuk/-/blob/master/src/openpgp.c#L930

**Comment by dengert** [comment]
@alex-nitrokey I have the Nitro Start you sent me few months ago is serial 00123355. Can you confirm if this has or does not have the above limit? I have it initialized, and hope to test the OpenSC changes using this token.

**Comment by dengert** [comment]
I cloned https://github.com/ya-isakov/OpenSC.git checked out branch ed25519. 
` ./pkcs11-tool --sign --id 3 -i /tmp/300.txt -o /tmp/300.sig` fails (330 byte file):
```
Outgoing APDU (260 bytes):
10 88 00 00 FF 30 31 32 33 34 35 36 37 38 39 0A .....0123456789.
30 31 32 33 34 35 36 37 38 39 0A 30 31 32 33 34 0123456789.01234
35 36 37 38 39 0A 30 31 32 33 34 35 36 37 38 39 56789.0123456789
0A 30 31 32 33 34 35 36 37 38 39 0A 30 31 32 33 .0123456789.0123
34 35 36 37 38 39 0A 30 31 32 33 34 35 36 37 38 456789.012345678
39 0A 30 31 32 33 34 35 36 37 38 39 0A 30 31 32 9.0123456789.012
33 34 35 36 37 38 39 0A 30 31 32 33 34 35 36 37 3456789.01234567
38 39 0A 30 31 32 33 34 35 36 37 38 39 0A 30 31 89.0123456789.01
32 33 34 35 36 37 38 39 0A 30 31 32 33 34 35 36 23456789.0123456
37 38 39 0A 30 31 32 33 34 35 36 37 38 39 0A 30 789.0123456789.0
31 32 33 34 35 36 37 38 39 0A 30 31 32 33 34 35 123456789.012345
36 37 38 39 0A 30 31 32 33 34 35 36 37 38 39 0A 6789.0123456789.
30 31 32 33 34 35 36 37 38 39 0A 30 31 32 33 34 0123456789.01234
35 36 37 38 39 0A 30 31 32 33 34 35 36 37 38 39 56789.0123456789
0A 30 31 32 33 34 35 36 37 38 39 0A 30 31 32 33 .0123456789.0123
34 35 36 37 38 39 0A 30 31 32 33 34 35 36 37 38 456789.012345678
39 0A 30 31                                                             9.01
P:54394; T:0x140084895127616 18:21:34.229 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:242:pcsc_internal_transmit: called
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:333:pcsc_transmit:
Incoming APDU (2 bytes):
90 00 ..
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/apdu.c:537:sc_transmit: returning with: 0 (Success)
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/apdu.c:515:sc_transmit: called
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/apdu.c:363:sc_single_transmit: called
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/apdu.c:367:sc_single_transmit: CLA:0, INS:88, P1:0, P2:0, data(75) 0x7fff1e064d2f
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:323:pcsc_transmit: reader 'Nitrokey Nitrokey Start (FSIJ-1.2.15-43123355) 00 00'
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:324:pcsc_transmit:
Outgoing APDU (81 bytes):
00 88 00 00 4B 32 33 34 35 36 37 38 39 0A 30 31 ....K23456789.01
32 33 34 35 36 37 38 39 0A 30 31 32 33 34 35 36 23456789.0123456
37 38 39 0A 30 31 32 33 34 35 36 37 38 39 0A 30 789.0123456789.0
31 32 33 34 35 36 37 38 39 0A 30 31 32 33 34 35 123456789.012345
36 37 38 39 0A 30 31 32 33 34 35 36 37 38 39 0A 6789.0123456789.
00                                              .
P:54394; T:0x140084895127616 18:21:34.231 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:242:pcsc_internal_transmit: called
P:54394; T:0x140084895127616 18:21:34.233 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:333:pcsc_transmit:
Incoming APDU (2 bytes):
69 85 i.
```

`./pkcs11-tool --sign --id 3 -i /tmp/300.txt -o /tmp/300.sig` (101 byte file) works:

```
Outgoing APDU (107 bytes):
00 88 00 00 65 31 32 33 34 35 36 37 38 39 0A 31 ....e123456789.1
32 33 34 35 36 37 38 39 0A 31 32 33 34 35 36 37 23456789.1234567
38 39 0A 31 32 33 34 35 36 37 38 39 0A 31 32 33 89.123456789.123
34 35 36 37 38 39 0A 31 32 33 34 35 36 37 38 39 456789.123456789
0A 31 32 33 34 35 36 37 38 39 0A 31 32 33 34 35 .123456789.12345
36 37 38 39 0A 31 32 33 34 35 36 37 38 39 0A 31 6789.123456789.1
32 33 34 35 36 37 38 39 0A 0A 00                23456789...
P:54339; T:0x140273119074368 18:08:35.730 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:242:pcsc_internal_transmit: called
P:54339; T:0x140273119074368 18:08:35.772 [opensc-pkcs11] ../../../OpenSC/src/libopensc/reader-pcsc.c:333:pcsc_transmit: 
Incoming APDU (66 bytes):
13 D6 43 72 6F 09 F4 DB 47 C8 DF F7 08 A9 93 E1 ..Cro...G.......
53 6C 95 21 2E 66 5B 48 69 AC 6E 06 7E DF 12 FF Sl.!.f[Hi.n.~...
51 1B F7 D6 2D 7A 80 C5 59 72 CD 39 B2 06 1C 07 Q...-z..Yr.9....
D6 1E B1 54 4B E4 01 36 50 51 9A E7 F5 06 E2 03 ...TK..6PQ......
90 00  
```
Looks like my NitroKey Start has older version of applet.

**Comment by dengert** [comment]
@ya-isakov I got my Yubikey 5 NFC initialized with Ed25519 sign and authenticate keys.
But in the Yubikey in the D0C1 blob sets the algorithm to `16 2b 06 02 04 02 da 47 0f 01 00` 
But it has an extra 00, which it added to the OID. 
To get this to work I did:
```
diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index 18cb2061..5a27a6da 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -95,6 +95,8 @@ static pgp_ec_curves_t        ec_curves_openpgp[] = {
        {{{1, 3, 36, 3, 3, 2, 8, 1, 1, 7, -1}}, 256}, /* brainpoolP256r1 */
        {{{1, 3, 36, 3, 3, 2, 8, 1, 1, 11, -1}}, 384}, /* brainpoolP384r1 */
        {{{1, 3, 36, 3, 3, 2, 8, 1, 1, 13, -1}}, 512}, /* brainpoolP512r1 */
+       {{{1, 3, 6, 1, 4, 1, 11591, 15, 1, -1}}, 256}, /* ed25519 for signatures => CKK_EC_EDWARDS */
+       {{{1, 3, 6, 1, 4, 1, 11591, 15, 1, 0, -1}}, 256}, /* ed25519 for signatures => CKK_EC_EDWARDS */
        {{{-1}}, 0} /* This entry must not be touched. */
 };
```
I added 2 OIDs one with the extra zero. Not sure if this is a Yubikey or gpg error.

**Comment by ya-isakov** [comment]
@dengert yes, I'm aware of this issue, I fixed it properly (the same as gnupg) in my #2286

**Comment by dengert** [comment]
Well I cloned you repro, and checked out  ed25519_fix. 
What branch has all you changes for Ed22519?

**Comment by ya-isakov** [comment]
The change from #2286 is in master branch of my repo, you can merge both branches, so you'll have all commits.

**Comment by ya-isakov** [comment]
Oh, and proper branch for ed25519 is `ed25519`, not `ed25519_fix`

**Comment by alex-nitrokey** [comment]
I guess, @szszszsz may can help here, as he is mainly responsible for the NK Starts firmware. I do not know that much about the firmware, I am afraid.

**Comment by dengert** [comment]
I picked the ed25519_fix branch because it had the 1.3.6.1.4.1.11591.15.1  OID but the ed25519 branch had the 1.3.101.112 OID.  

I used gpg (GnuPG) 2.2.19 to initialize the tokens as suggested for Yibikey and Nitrokey and it appears write the 1.3.6.1.4.1.11591.15.1 to the DOs.

OIDs are usually allocated by a research organization while things are being developed or first implemented. Then if things like EdDSA become standards, then new OIDs may are also allocated.    

There are tokens and gpg code in distribution using 1.3.6.1.4.1.11591.15.1. So the code really needs to accept either when reading, and if not sure which to write, use the 1.3.101.112 when writing. A closer look at RFC 8410 and 8420 and any x509 is also needed. 

Note pkcs15-pubkey.c also has a mapping for 1.3.6.1.4.1.11591.15.1. This may also need updating.  

http://www.oid-info.com/ can be very helpful. It lists 1,534,967 OIDs. 

http://www.oid-info.com/cgi-bin/display?oid=1.3.6.1.4.1.11591.15.1&submit=Display&action=display
http://www.oid-info.com/cgi-bin/display?oid=1.3.101.112&submit=Display&action=display

http://www.oid-info.com/cgi-bin/display?oid=1.3.6.1.4.1.11591&submit=Tree+display

shows 11591 is  `Free Software Foundation, "GNU" Project`


ECDSA has many curves, represented by namedCurve that map to OIDs. It looks like EdDSA dropped the concept of named curves, and just uses OIDs. 

EdDSA actually has 2 curves: Ed25519 and Ed448. I don't know if any tokens support Ed448, but any code changes should anticipate there may be more ED curves and there will be tokens with Ed448. 
 
./framework-pkcs15.c has a couple of TODO comments about edwards25519" or "edwards448" and 
"curve25519" or "curve448" which should also be looked at.

**Comment by ya-isakov** [comment]
@dengert The whole point of ed25519 branch was to have a RFC8410 compatibility for pkcs11. That's why there are proper OIDs, from this RFC, in ed25519 branch. `1.3.6.1.4.1.11591.15.1` is OpenPGP oid, and without my fixes from ed25519 branch (i.e. #2293), every public key will not be parsable by x509 (could be easily checked by `pkcs15-tool --read-public-key 01 | openssl pkey -pubin -text`, I have it in PR description)

**Comment by ya-isakov** [comment]
So, for card-openpgp.c it must be 1.3.6.1.4.1.11591.15.1, but for pkcs15-algo it must be RFC8410-defined OID

**Comment by ya-isakov** [comment]
This branch is passing p11test for me, as without these changes, openssl in these tests cannot parse OIDs from public keys (as 1.3.6.1.4.1.11591.15.1 are not defined in any X509-related RFCs)

**Comment by ya-isakov** [comment]
Same thing for ECDSA - it's defined that OID shoud be 1.2.840.10045.2.1, followed by curve name, and in card-openpgp there is no such OID, as OpenPGP uses it's own OIDs for every curve, instead of name

**Comment by ya-isakov** [comment]
To conclude - toghether #2286 and #2293 allows Yubikey to be fully RFC-conformant with RFC8410, RFC8032, for sign using EdDSA. Derive also works (including p11test), but with last commit from #2287.

**Comment by ya-isakov** [comment]
Without my changes, p11 test will show similar output as in e.g. #1960, but it's just ignoring tests with errors. You can compile p11test without -DNDEBUG in Makefile, on e.g. master, and see that there are plenty of errors.

**Comment by dengert** [comment]
> @dengert The whole point of ed25519 branch was to have a RFC8410 compatibility for pkcs11. That's why there are proper OIDs, from this RFC, in ed25519 branch. 1.3.6.1.4.1.11591.15.1 is OpenPGP oid, and without my fixes from ed25519 branch (i.e. #2293), every public key will not be parsable by x509 (could be easily checked by pkcs15-tool --read-public-key 01 | openssl pkey -pubin -text, I have it in PR description)

Based on the other comments after the above, you are doing the correct thing. OpenPGP or at least GnuPG is inserting the  1.3.6.1.4.1.11591.15.1 on the tokens. So you must be able to handle this when reading "OpenPGP" data.  

How did you initialize you Yubikey? What OID is being read? 

I have downloaded GnuPG 3.0, and ./common/openpgp-oid.c has:
```
 49   { "Curve25519", "1.3.6.1.4.1.3029.1.5.1", 255, "cv25519", PUBKEY_ALGO_ECDH },
 50   { "Ed25519",    "1.3.6.1.4.1.11591.15.1", 255, "ed25519", PUBKEY_ALGO_EDDSA },
 51   { "X448",       "1.3.101.111",            448, "cv448",   PUBKEY_ALGO_ECDH },
 52   { "Ed448",      "1.3.101.113",            456, "ed448",   PUBKEY_ALGO_EDDSA },
```

This implies they are using "GnuPG" OID for some curves, and "Thawte Consulting" for others. This looks historic to me, in that they may have implementing EdDSA ed25519 internally and used one of their OIDs, and continue to use it, while for ed448 they are using the "official" "RFC" OID.

These are the kinds of issues you are running into. The scope of the OID may be limited, and two OID s may be used for the same thing.

**Comment by ya-isakov** [comment]
Yes, these are correct OpenPGP/Card OIDs (for 25519), and they're on card, that's why I have these in my #2286 (master branch of my fork). I have no idea, what they should be for 448. Probably, for these, they chosen to follow RFC8410 (nice, then).

**Comment by ya-isakov** [comment]
RFC4880-bis-10 has no OIDs for curves in 448 series, so, maybe gnupg will be changed, after RFC will be finalized (2.3.0 is not stable, anyway)

**Comment by frankmorgner** [comment]
FYI, I've stumbled across the same issue in [OpenSCToken](https://github.com/frankmorgner/OpenSCToken/pull/40) and I have implemented a quick fix in [a seperate branch](https://github.com/OpenSC/OpenSC/tree/new_algs). The change for pkcs15-sec.c is not very sophisticated and should be reworked, but it (hopefully) fixes the problem for now. The problem in mechanism.c still remains, however.

**Comment by frankmorgner** [comment]
I've added a patch to allow an arbitrary length for signing in PKCS#11. So far, the CI tests look good. You may want to have a look...

**Comment by ya-isakov** [comment]
@frankmorgner Testing it, thank you!

**Comment by ya-isakov** [comment]
@frankmorgner I successfully signed file of length 3053, so, it seems that it is working.
